client/db/src/lib.rs

```diff
@@ -474,13 +478,112 @@ fn cache_header<Hash: std::cmp::Eq + std::hash::Hash, Header>(
 	}
 }
 
+struct PinnedBlockCacheEntry<Block: BlockT> {
+	ref_count: u32,
+	pub justifications: Option<Option<Justifications>>,
+	pub body: Option<Option<Vec<Block::Extrinsic>>>,
+}
+
+impl<Block: BlockT> Default for PinnedBlockCacheEntry<Block> {
+	fn default() -> Self {
+		Self { ref_count: 0, justifications: None, body: None }
+	}
+}
+
+impl<Block: BlockT> PinnedBlockCacheEntry<Block> {
+	pub fn decrease_ref(&mut self) {
+		self.ref_count = self.ref_count.saturating_sub(1);
+	}
+
+	pub fn increase_ref(&mut self) {
+		self.ref_count = self.ref_count.saturating_add(1);
+	}
+
+	pub fn has_no_references(&self) -> bool {
+		self.ref_count == 0
+	}
+}
+
+struct PinnedBlockCache<Block: BlockT> {
+	cache: RwLock<LruCache<Block::Hash, PinnedBlockCacheEntry<Block>>>,
+}
+
+impl<Block: BlockT> PinnedBlockCache<Block> {
+	pub fn new() -> Self {
+		Self { cache: LruCache::new(NonZeroUsize::new(PINNING_CACHE_SIZE).unwrap()).into() }
```

Comment:
(*shameless plug incoming*) Another option would be to use `LruMap::new(ByLength(PINNING_CACHE_SIZE))` from `schnellru` which doesn't require an `unwrap`... (: You could also maybe copy-paste the `ByLength` limiter impl and implement the `on_removed` and put a log there to easily print out which exact element was removed, and you wouldn't have to manually track when any entries are removed since the map automatically calls that callback when an entry is evicted.

But it's up to you.

Reply:
Switched over and moved the removal logging to the limiter.